cpp/nacjd.cc

```diff
@@ -0,0 +1,371 @@
+        return false;
+    }
+     // Check for rate limiting and error status codes:
+    const HttpHeader http_header(downloader.getMessageHeader());
+    if (http_header.getStatusCode() == 429)
+        LOG_ERROR("we got rate limited!");
+    else if (http_header.getStatusCode() != 200) {
+        LOG_WARNING("NACJD returned HTTP status code " + std::to_string(http_header.getStatusCode()) + "! for NACJD id: " + id);
+        return false;
+    }
+    const std::string &json_document(downloader.getMessageBody());
+    std::shared_ptr<JSON::JSONNode> full_tree;
+    JSON::Parser parser(json_document);
+    if (not parser.parse(&full_tree)) {
+        LOG_ERROR("failed to parse JSON (" + parser.getErrorMessage() + "), download URL was: " + DOWNLOAD_URL);
+        return false;
+    }
+    const std::shared_ptr<const JSON::ObjectNode> top_node(JSON::JSONNode::CastToObjectNodeOrDie("full_tree", full_tree));
+    if (use_separator)
+        json_new_titles << "," << std::endl;
+    json_new_titles << top_node->toString() << std::endl;
+    return true;
+}
+
+
+void extractExistingIDsFromMarc(MARC::Reader *const marc_reader, std::set<std::string> &parsed_marc_ids) {
+    while (MARC::Record record = marc_reader->read()) {
+        std::string ppn = record.getControlNumber();
+        std::string id_035 = record.getFirstSubfieldValue("035", 'a');
+        if (StringUtil::Contains(ppn, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &ppn);
+            StringUtil::TrimWhite(&ppn);
+            parsed_marc_ids.emplace(ppn);
+        }
+        if (StringUtil::Contains(id_035, "[ICPSR]")) {
+            StringUtil::ReplaceString("[ICPSR]", "", &id_035);
+            StringUtil::TrimWhite(&id_035);
+            parsed_marc_ids.emplace(id_035);
+        }
+    }
+}
+
+
+void extractIDsFromWebsite(std::set<std::string> &parsed_marc_ids, unsigned * const number_of_new_ids) {
+    const std::string DOWNLOAD_URL = "https://www.icpsr.umich.edu/web/NACJD/search/studies?start=0&ARCHIVE=NACJD&PUBLISH_STATUS=PUBLISHED&sort=DATEUPDATED%20desc&rows=9000";
+    if (FileUtil::Exists(nacjd_titles))
+        FileUtil::DeleteFile(nacjd_titles);
+    
+    if (not Download(DOWNLOAD_URL, nacjd_titles, timeout * 1000))
+        LOG_ERROR("Could not download website with nacjd ids.");
+    std::ifstream file(nacjd_titles);
+    if (file)
+        std::for_each(std::istream_iterator<char>(file),
+                      std::istream_iterator<char>(),
+                      handleChar);
+    else {
+        LOG_ERROR("couldn't open file: " + nacjd_titles);
+    }
+    
+    if (FileUtil::Exists(nacjd_new_titles_json))
+        FileUtil::DeleteFile(nacjd_new_titles_json);
+    std::ofstream json_new_titles;
+    json_new_titles.open(nacjd_new_titles_json);
+    json_new_titles << "{ \"nacjd\" : [ " << std::endl;
+    bool first = true;
+    for (std::string id : ids_website) {
```

Comment:
Please make id a const reference to avoid inefficient creation of  string variables.  Also, I would prefer auto.